### PR TITLE
Fix covariant returns validation

### LINKED ISSUE #41853: R2R covariant returns test failures
R2R test failures in apparently every configuration/architecture.

```
    Loader\classloader\MethodImpl\CovariantReturns\UnitTest\OverrideMoreDerivedReturn\OverrideMoreDerivedReturn.cmd [FAIL]
      
      Assert failure(PID 2908 [0x00000b5c], Thread: 5808 [0x16b0]): HasStableEntryPoint()
      
      CROSSGEN! MethodTable::SetSlot + 0x116 (0x00007ff6`4c19d2ea)
      CROSSGEN! ClassLoader::ValidateMethodsWithCovariantReturnTypes + 0x650 (0x00007ff6`4c1a358c)
      CROSSGEN! PushFinalLevels + 0x23 (0x00007ff6`4c1bba67)
      CROSSGEN! ClassLoader::LoadTypeHandleForTypeKey + 0x197 (0x00007ff6`4c1b9c0f)
      CROSSGEN! ClassLoader::LoadConstructedTypeThrowing + 0x200 (0x00007ff6`4c1b8c84)
      CROSSGEN! ClassLoader::LoadGenericInstantiationThrowing + 0x1ED (0x00007ff6`4c1b8f35)
      CROSSGEN! SigPointer::GetTypeHandleThrowing + 0xB3F (0x00007ff6`4c1c8323)
      CROSSGEN! LoadTypeSpecHelper + 0xC4 (0x00007ff6`4c1ebfd4)
      CROSSGEN! Module::ExpandAll + 0x6D5 (0x00007ff6`4c1e46ed)
      CROSSGEN! CEEPreloader::Preload + 0x7A (0x00007ff6`4c173bfe)
          File: F:\workspace\_work\1\s\src\coreclr\src\vm\method.hpp Line: 248
          Image: C:\h\w\A31108E3\p\crossgen.exe
```

Starting with: https://dev.azure.com/dnceng/public/_build/results?buildId=797990&view=ms.vss-test-web.build-test-results-tab

Failures include:
```
    Loader\classloader\MethodImpl\CovariantReturns\UnitTest\OverrideMoreDerivedReturn\OverrideMoreDerivedReturn.cmd [FAIL]
    Loader\classloader\MethodImpl\CovariantReturns\UnitTest\UnitTestDelegates\UnitTestDelegates.cmd [FAIL]
    Loader\classloader\MethodImpl\CovariantReturns\UnitTest\UnitTest_GVM\UnitTest_GVM.cmd [FAIL]
    Loader\classloader\MethodImpl\CovariantReturns\Interfaces\UnitTest\UnitTest.cmd [FAIL]
    Loader\classloader\MethodImpl\CovariantReturns\UnitTest\UnitTest\UnitTest.cmd [FAIL]
```

Likely candidate PR: https://github.com/dotnet/runtime/pull/41703

@davidwrighton @janvorli PTAL

### PR DESCRIPTION
**Pull request by janvorli**
Recent fix to a problem with covariant returns validation has introduced
another problem related to code versioning. The validation method is now
called too late in the type load process which causes problem in
crossgening or in case code versioning is disabled.

This change fixes it by splitting the return type verification and
mutating changes like vtable slot updates into two separate phases, each
called at a proper time.

Close #41853

### CODE REVIEW

File: src/coreclr/src/vm/class.cpp [hunk 1]
```diff
@@ -1265,6 +1281,11 @@ void ClassLoader::ValidateMethodsWithCovariantReturnTypes(MethodTable* pMT)
             if (pMD == pParentMD)
                 continue;
 
+            // If the bit is not set on this method, but we reach here because it's been set on the method at the same slot on
+            // the base type, set the bit for the current method to ensure any future overriding method down the chain gets checked.
+            if (!pMD->RequiresCovariantReturnTypeChecking())
+                pMD->SetRequiresCovariantReturnTypeChecking();
```
Comment (davidwrighton) [inline src/coreclr/src/vm/class.cpp:1287]: This is setting RequiresCovariantReturnTypeChecking too often. In particular, we should only force set the bit here if the parenteMD RequiresCovariantReturnTypeChecking.

**Review by davidwrighton — CHANGES_REQUESTED** [review]
You need to avoid setting RequiresCovariantTypeChecking so often. Covariant type checking is expensive, and while setting the bit extra times isn't a correctness problem it will show up in a profile of the application.

Also, last night I said the assert that was firing looked like there may be a concern caused by misbehavior of MethodTableBuilder::ChangesImplementationOfVirtualSlot in the presence of this feature. However, that assert is pretty fragile and incomplete. These changes make it so that the assert will not fire, but I do not have confidence that the vtable chunk sharing logic may be misbehaving. Have you had a chance to look into that?

**Comment by janvorli** [comment]
> Have you had a chance to look into that?

I did and it seems to me there is really a problem there and that the MethodTableBuilder::ChangesImplementationOfVirtualSlot needs to be modified to consider the attribute. However, I'd leave that for another PR since for now, this issue is blocking outerloop, so it needs to be fixed asap.

**Comment by janvorli** [comment]
> You need to avoid setting RequiresCovariantTypeChecking so often.

Right, the check for the parent MD has somehow fallen out of my latest version by mistake.

**Review by davidwrighton — APPROVED** [review]

**Pull request merged**